### PR TITLE
jax.numpy: better docstring for append function

### PR DESCRIPTION
**Pull request by selamw1**
New docstring and JAX-specific code examples are addd for jax.numpy.append.
Part of [21461](https://github.com/google/jax/issues/21461).

### CODE REVIEW

File: jax/_src/numpy/lax_numpy.py [hunk 1]
```diff
@@ -3955,12 +3955,49 @@ def trim_zeros_tol(filt, tol, trim='fb'):
...
     arr: ArrayLike, values: ArrayLike, axis: int | None = None
 ) -> Array:
+  """Return a new array with values appended to the end of the original array.
+
+  JAX implementation of :func:`numpy.append`.
+
+  Args:
+    arr: original array.
+    values: values to be appended to the array.
+    axis: axis along which to append values.
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:3969]: Maybe mention that if `axis=None`, both `arr` and `values` will be flattened before appending.

File: jax/_src/numpy/lax_numpy.py [hunk 2]
```diff
@@ -3955,12 +3955,49 @@ def trim_zeros_tol(filt, tol, trim='fb'):
...
 def append(
     arr: ArrayLike, values: ArrayLike, axis: int | None = None
 ) -> Array:
+  """Return a new array with values appended to the end of the original array.
+
+  JAX implementation of :func:`numpy.append`.
+
+  Args:
+    arr: original array.
+    values: values to be appended to the array.
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:3968]: We should mention that `values` must have the same number of dimensions as `arr`, and all dimensions must match except in the specified axis.

File: jax/_src/numpy/lax_numpy.py [hunk 3]
```diff
@@ -3955,12 +3955,49 @@ def trim_zeros_tol(filt, tol, trim='fb'):
...
+
+  Examples:
+    >>> a = jnp.array([1, 2, 3])
+    >>> b = jnp.array([4, 5, 6])
+    >>> jnp.append(a, b)
+    Array([1, 2, 3, 4, 5, 6], dtype=int32)
+
+    Appending along a specific axis:
+
+    >>> a = jnp.array([[1, 2], [3, 4]])
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:3986]: For the sake of example, I think it's easier for the reader to visualize the semantics if you define the array like this:
```python
    >>> a = jnp.array([[1, 2],
    ...                [3, 4]])
```
Same below.

**Review by jakevdp — CHANGES_REQUESTED** [review]

File: jax/_src/test_util.py [hunk 4]
```diff
@@ -178,11 +178,26 @@ def check_eq(xs, ys, err_msg=''):
 
 
 @contextmanager
-def capture_stdout() -> Generator[Callable[[], str], None, None]:
-  with unittest.mock.patch('sys.stdout', new_callable=io.StringIO) as fp:
-    def _read() -> str:
-      return fp.getvalue()
-    yield _read
+def capture_stdout() -> Generator[Callable[[], str | None], None, None]:
```
Comment (jakevdp) [inline jax/_src/test_util.py:181]: Unrelated change?

File: jax/_src/numpy/lax_numpy.py [hunk 5]
```diff
@@ -3965,8 +3965,8 @@ def append(
 
   Args:
     arr: original array.
-    values: values to be appended to the array.
-    axis: axis along which to append values.
+    values: values to be appended to the array. The ``values`` must have the same number of dimensions as ``arr``, and all dimensions must match except in the specified axis.
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:3968]: For docstrings, we should manually break lines after ~80ish characters, otherwise the formatting is very poor when displayed in a terminal. See other docstrings in this file for examples of how to indent this.

**Review by jakevdp — CHANGES_REQUESTED** [review]
It looks like you may have rebased against an older version of the main branch, which caused commits from the newer version of the main branch to be considered part of your PR. The best fix would be to rebase again against the most recent main branch.

File: jax/_src/numpy/lax_numpy.py [hunk 6]
```diff
@@ -3955,12 +3955,54 @@ def trim_zeros_tol(filt, tol, trim='fb'):
...
+
+  JAX implementation of :func:`numpy.append`.
+
+  Args:
+    arr: original array.
+    values: values to be appended to the array. The ``values`` must have 
+    the same number of dimensions as ``arr``, and all dimensions must 
+    match except in the specified axis.
+    axis: axis along which to append values. If None (default), both ``arr`` 
+    and ``values`` will be flattened before appending.
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:3972]: Please add two spaces of indentation to lines 3969, 3970, and 3972.

You can look at other docstrings in this file to see correct indentation of argument lists.

File: jax/_src/numpy/lax_numpy.py [hunk 7]
```diff
@@ -3955,12 +3955,54 @@ def trim_zeros_tol(filt, tol, trim='fb'):
...
+
+    >>> a = jnp.array([[1, 2],
+    ...                [3, 4]])
+    >>> b = jnp.array([[5, 6]])
+    >>> jnp.append(a, b, axis=0)
+    Array([[1, 2],
+           [3, 4]],
+           [5, 6]], dtype=int32)
+
+    Appending with a different shape along axis:
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:3997]: I don't understand what this means. Different shape than what? Maybe "Appending along a trailing axis" would be a better description?

**Review by jakevdp — CHANGES_REQUESTED** [review]

File: jax/_src/numpy/lax_numpy.py (hunk 7 above)
Comment (selamw1) [reply jax/_src/numpy/lax_numpy.py]: yes, its modified and emphasizes the action of adding elements along the last dimension.

**Review by jakevdp — COMMENTED** [review]
Looks great! Last thing we need is to squash all the changes into a single commit, and then we can merge (see https://jax.readthedocs.io/en/latest/contributing.html#single-change-commits-and-pull-requests) Thanks!

**Comment by selamw1** [comment]
@jakevdp all commits squashed into one.

**Review by jakevdp — APPROVED** [review]
Looks good, thanks!
